### PR TITLE
App Submission: PeerSwap

### PR DESCRIPTION
**Pull request by Impa10r**
# App Submission

### App name
PeerSwap

### Version
1.4.7

### One line description of the app
Balance your lightning channels with Liquid BTC

### Summary of the app
PeerSwap enables Lightning Network nodes to balance their channels by facilitating atomic swaps with direct peers. PeerSwap enhances decentralization of the Lightning Network by enabling all nodes to be their own swap provider. No centralized coordinator, no 3rd party rent collector, and lowest cost channel balancing means small nodes can better compete with large nodes.

### Developer name
PeerSwap Developers

### Developer website
https://peerswap.dev

### Source code link
https://github.com/Impa10r/peerswap-web

### Support link
https://discord.com/invite/wpNv3PG8G2

### Requires
- [x] Elements Core
- [x] Lightning Node

### 256x256 SVG icon
[Submit an icon](https://svgur.com/s/11kY)

### Gallery images

A powerful companion for your node.

image

Cheap channel rebalancing with Liquid.

image

Cost effective Liquid sourcing with Peg-In.

image

### I have tested my app on:
- [x] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [x] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

File: peerswap/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      ELEMENTS_USER: elements
+      ELEMENTS_PORT: $APP_ELEMENTS_NODE_RPC_PORT
+      ELEMENTS_HOST: http://elements_node_1
+      ELEMENTS_WALLET: peerswap
+      ELEMENTS_FOLDER: /home/elements/.elements
+      ELEMENTS_FOLDER_MAPPED: /root/.elements
+    volumes:
+      - ${APP_DATA_DIR}/data:/root/.peerswap
+      - ${APP_LIGHTNING_NODE_DATA_DIR}:/root/.lnd:ro
+      - /home/umbrel/umbrel/app-data/elements/data:/root/.elements:ro
```
Comment (smolgrrr) [inline peerswap/docker-compose.yml:29]: Hi @Impa10r thanks for the submission- excited to try this approach for rebalancing my LND node.

Unfortunately I wasn't able to get this running on an Umbrel instance just yet.
I've made some small changes, but I can see there are some issues importing variables here- will get back to you shortly with the best approach for resolving this.

Comment (Impa10r) [reply peerswap/docker-compose.yml:29]: @smolgrrr Hello. Anything I can help you with?

Comment (thool2b3) [reply peerswap/docker-compose.yml:29]: @smolgrrr 
Hello, I am very excited to see that Peerswap Web is going to be integrated in the official Umbrel App store. 
Could you please give us an update on your progress.
Thanks a lot.

**Comment by Impa10r** [comment]
@smolgrrr Hello. I updated the app to run as non-root, works fine on Umbrel 0.5.4.

File: peerswap/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      ELEMENTS_USER: elements
+      ELEMENTS_PORT: $APP_ELEMENTS_NODE_RPC_PORT
+      ELEMENTS_HOST: http://elements_node_1
+      ELEMENTS_WALLET: peerswap
+      ELEMENTS_FOLDER: /home/elements/.elements
+      ELEMENTS_FOLDER_MAPPED: /home/peerswap/.elements
+    volumes:
+      - ${APP_DATA_DIR}/data:/home/peerswap/.peerswap
+      - ${APP_LIGHTNING_NODE_DATA_DIR}:/home/peerswap/.lnd:ro
+      - /home/umbrel/umbrel/app-data/elements/data:/home/peerswap/.elements:ro
```
Comment (nmfretz) [inline peerswap/docker-compose.yml:30]: To make sure this works for all users, including those who are running pre-1.0 umbrelOS on custom linux installs, we should export the elements core data directory as an environment variable from the PeerSwaps exports.sh, and define it as a relative path from the script.

You would do something like this (untested):
```bash
ELEMENTS_DATA_DIR="$(readlink -f $(dirname "${BASH_SOURCE[0]}")/..)/elements/data"
```

Right now, a custom linux user on umbrelOS 0.5.x might have their umbrel install at a path other than `/home/umbrel/umbrel`. For example, it's possible that their elements data directory would be at something like `/home/skye/installs/umbrel/app-data/elements/data`. If PeerSwap hardcodes `/home/umbrel/umbrel/app-data/elements/data` then this user wouldn't be able to use the app.

File: peerswap/exports.sh
Comment (nmfretz) [inline peerswap/exports.sh:1]: This is the file you can export the ELEMENTS_DATA_DIR environment variable from.

File: peerswap/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+version: "3.7"
+
+services:
+  app_proxy:
+    environment:
+      APP_HOST: peerswap_web_1
+      APP_PORT: 1984
```
Comment (nmfretz) [inline peerswap/docker-compose.yml:7]: solid port choice 😉

**Review by nmfretz — COMMENTED** [review]
Thanks @Impa10r! Great work building and submitting PeerSwap. I am currently syncing Elements Core to properly test.

In the meantime, I have reviewed and left a couple minor suggestions.

We can help get gallery images and app icon finalized for the app store:
- Do you have a preference on which 3 to use of the 6 you submitted?
- Do you have any ideas for text in each gallery image?

**Comment by Impa10r** [comment]
I am currently struggling to install Umbrel 1.1 testing environment in VMWare. My old Umbrel 0.5.4 test setup ran out of disk space. Will attend this conversation when able.

File: peerswap/exports.sh
Comment (Impa10r) [reply peerswap/exports.sh:1]: @nmfretz Thanks for reviewing! I implemented your suggestions, tested with Umbrel 1.1.2 and updated the screenshots in the Gallery Images.

**Comment by Impa10r** [comment]
@nmfretz I just noticed that in Umbrel 1.1.2 RBAC is not allowing my app to dial mempool.space, github.com and getblock.io. Can this be circumvented via a tor proxy? Any trick we can use to provide it as an env variable?

**Comment by Impa10r** [comment]
@nmfretz I see trouble connecting to elements with Umbrel 1.1.2. Will investigate.

**Comment by Impa10r** [comment]
@nmfretz Fixed it in v1.3.6

**Comment by Impa10r** [comment]
> I noticed one thing you may want to clarify in the UI and/or the app description at some point. 

Yes, thanks, will make this error message more clear in the next update

**Comment by Impa10r** [comment]
Hi @nmfretz, how is it going? In my next release v1.5.0 I enabled HTTPS option for baremetal users, but am struggling to do that with Umbrel. If the option is enabled, port 1984 starts redirecting all calls to HTTPS port 1985. Is it possible to have two app_proxy processes for one app? Or two ports open for app_proxy?

**Comment by nmfretz** [comment]
Hey Vlad, the gallery assets are now done, and we can go live!
icon

**Comment by nmfretz** [comment]
> Hi @nmfretz, how is it going? In my next release v1.5.0 I offer HTTPS connections, but am struggling to do that with Umbrel. If the option is enabled, HTTP port 1984 starts redirecting all calls to HTTPS port 1985. Is it possible to have two app_proxy processes for one app? Or two ports open for app_proxy?

Ah, there is no way right now to have two ports open for the app_proxy. But even more importantly, https is not implemented right now in umbrelOS. We are on the verge of releasing it, so as soon as that is done we can update PeerSwap to work over https.

Are you okay with releasing v1.4.7 now and then updating when https is available?

**Comment by Impa10r** [comment]
> > Hi @nmfretz, how is it going? In my next release v1.5.0 I offer HTTPS connections, but am struggling to do that with Umbrel. If the option is enabled, HTTP port 1984 starts redirecting all calls to HTTPS port 1985. Is it possible to have two app_proxy processes for one app? Or two ports open for app_proxy?
> 
> Ah, there is no way right now to have two ports open for the app_proxy. But even more importantly, https is not implemented right now in umbrelOS. We are on the verge of releasing it, so as soon as that is done we can update PeerSwap to work over https.
> 
> Are you okay with releasing v1.4.7 now and then updating when https is available?

Yes

**Comment by nmfretz** [comment]
Excellent, thanks a ton for all your efforts bringing PeerSwap to Umbrel @Impa10r! Going live to the app store now ⚡️.

**Pull request merged**